### ISSUE TITLE
Argument type coercion mixed in static factory method

### ISSUE THREAD

**Issue by akondas** [vimeo/psalm#2283]
Using template and static factory method gives `MixedArgumentTypeCoercion`:

Link: https://psalm.dev/r/b039de88aa

Code:
```php
<?php
 
/**
 * @template T
 */
class Lazy {
  
  /**
   * @var T
   */
  private $value;
  
  /**
   * @param T $value
   */
  private function __construct($value)
  {
    $this->value = $value;
  }
  
  /**
   * @param T $value
   * @return Lazy<T>
   */
  public static function of($value)
  {
    return new self($value);
  }
}

class Foo {
  /**
   * @param Lazy<int> $lazy
   */
  public function __construct($lazy)
  {

  }
}

$lazy = Lazy::of(1);

new Foo($lazy);
```

Error:
```
INFO: MixedArgumentTypeCoercion - 43:9 - Argument 1 of Foo::__construct expects Lazy<int>, parent type Lazy<mixed> provided
```

Expected: no error

**Comment by muglug** [comment]
Just scope the template for that function properly: https://psalm.dev/r/170a3bf273
